### PR TITLE
Add secondary index for database server lookup by database name

### PR DESCRIPTION
**Pull request by wethreetrees**
Replaces the full scan in `IsMFARequiredRequest_Database`, which fetched all database servers and filtered in-memory, with a range scan over a secondary index keyed by database name. 

Adds a new function, `RangeDatabaseServersWithName`, to the cache and PresenceInternal interfaces that returns an iterator over database servers matching a given database name.

Also adds a generic test helper and benchmark to verify behavior of this change and the upcoming App and Kubernetes server changes.

Changelog: Improved performance and reduced resource usage of the auth service for clusters with large numbers of registered databases with per-session MFA enabled.

### Test Environment

Teleport Cloud tenant with at least one enrolled database and a role that requires per-session MFA for database access.

### Test Cases
- [x] Connect to a database where the role requires per-session MFA and confirm MFA prompt appears and connection succeeds after verification
- [x] Connect to a database where no per-session MFA is required and confirm connection succeeds without an MFA prompt
- [x] Unregister a previously registered database, attempt to connect, and confirm a not found error is returned

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -7989,17 +7989,15 @@ func (a *Server) isMFARequired(ctx context.Context, scopedCtx *authz.ScopedConte
 		if t.Database.ServiceName == "" {
 			return nil, trace.BadParameter("missing ServiceName field in a database-only UserCertsRequest")
 		}
-		servers, err := a.GetDatabaseServers(ctx, apidefaults.Namespace)
+		servers, err := a.GetDatabaseServersByDatabaseName(ctx, apidefaults.Namespace, t.Database.ServiceName)
```
[the remark this answers is not in the document]
Comment (wethreetrees) [reply lib/auth/auth.go]: Good catch. I made a mistake and added `GetDatabaseServersByDatabasename` to the `ReadProxyAccessPoint` interface instead of the `Cache` interface... Didn't notice immediately since the presence service changes also reduced memory usage and there were no build errors or test failures. 😅

I've removed the func from `ReadProxyAccessPoint` and added it to `Cache`.

File: lib/cache/database.go [hunk 2]
```diff
@@ -224,6 +235,38 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
 }
 
+// GetDatabaseServersByDatabaseName returns all registered database proxy servers for a given database name.
+func (c *Cache) GetDatabaseServersByDatabaseName(ctx context.Context, namespace, databaseName string, opts ...services.MarshalOption) ([]types.DatabaseServer, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetDatabaseServersByDatabaseName")
+	defer span.End()
+
+	if namespace == "" {
+		return nil, trace.BadParameter("namespace cannot be empty")
+	}
```
Comment (wethreetrees) [inline lib/cache/database.go]: Wasn't sure how to handle `namespace` for new functions. I added support to match other functions, but maybe this should just be removed in all the changes in the PR?

File: lib/cache/database.go [hunk 3]
```diff
@@ -180,6 +182,9 @@ func newDatabaseServerCollection(p services.Presence, w types.WatchKind) (*colle
 				databaseServerNameIndex: func(u types.DatabaseServer) string {
 					return u.GetHostID() + "/" + u.GetName()
 				},
+				databaseServerDatabaseNameIndex: func(u types.DatabaseServer) string {
+					return u.GetDatabase().GetName() + "/" + u.GetHostID() + "/" + u.GetName()
+				},
```
Comment (espadolini) [inline lib/cache/database.go]: As per https://github.com/gravitational/teleport/blob/master/rfd/0153-resource-guidelines.md#sorting-support , new non-backend indices should use `rsc.io/ordered`.

File: lib/cache/database.go (hunk 2 above)
Comment (espadolini) [reply lib/cache/database.go]: Definitely remove it, it won't contribute to searching anyway.

File: lib/cache/database.go [hunk 4]
```diff
@@ -224,6 +229,38 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
 	return out, nil
 }
 
+// GetDatabaseServersByDatabaseName returns all registered database proxy servers for a given database name.
+func (c *Cache) GetDatabaseServersByDatabaseName(ctx context.Context, namespace, databaseName string, opts ...services.MarshalOption) ([]types.DatabaseServer, error) {
```
Comment (espadolini) [inline lib/cache/database.go]: This should either be an iterator or a list operation with pagination, every time we made an assumption about some amount of items being "reasonable" has resulted in a problem later on.

File: lib/services/presence.go [hunk 5]
```diff
@@ -155,6 +155,8 @@ type Presence interface {
 
 	// GetDatabaseServers returns all registered database proxy servers.
 	GetDatabaseServers(context.Context, string, ...MarshalOption) ([]types.DatabaseServer, error)
+	// GetDatabaseServersByDatabaseName returns all registered database proxy servers for a given database name.
+	GetDatabaseServersByDatabaseName(ctx context.Context, namespace, databaseName string, opts ...MarshalOption) ([]types.DatabaseServer, error)
```
Comment (espadolini) [inline lib/services/presence.go]: I don't think we should be taking marshal options from callers.

File: lib/services/presence.go
Comment (espadolini) [inline lib/services/presence.go:204]: Move the method to PresenceInternal so we don't have to stub it out with an unconditional error in remote clients.

Comment (wethreetrees) [reply lib/services/presence.go:204]: That seems a lot better, thanks!

File: lib/cache/database.go (hunk 3 above)
Comment (wethreetrees) [reply lib/cache/database.go]: Updated the new index to use `rsc.io/ordered`

File: lib/cache/database.go (hunk 2 above)
Comment (wethreetrees) [reply lib/cache/database.go]: Done

File: lib/cache/database.go (hunk 4 above)
Comment (wethreetrees) [reply lib/cache/database.go]: Updated to return an iterator

File: lib/services/presence.go (hunk 5 above)
Comment (wethreetrees) [reply lib/services/presence.go]: Removed `opts`

File: lib/services/presence.go
Comment (wethreetrees) [reply lib/services/presence.go:204]: Moved the method into PresenceInternal

File: lib/auth/authclient/api.go [hunk 6]
```diff
@@ -1308,6 +1308,9 @@ type Cache interface {
 	// GetDatabaseServers returns all registered database proxy servers.
 	GetDatabaseServers(ctx context.Context, namespace string, opts ...services.MarshalOption) ([]types.DatabaseServer, error)
 
+	// GetDatabaseServersByDatabaseName returns an iterator over database proxy servers for a given database name.
+	GetDatabaseServersByDatabaseName(ctx context.Context, databaseName string) iter.Seq2[types.DatabaseServer, error]
```
Comment (espadolini) [inline lib/auth/authclient/api.go]: Following the naming scheme we've been using this should be `RangeDatabaseServersByDatabaseName` I think, although "by name" might suggest ordering by database name rather than filtering. Maybe `RangeDatabaseServersWithName`?

File: lib/cache/database.go [hunk 7]
```diff
@@ -224,6 +235,65 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
+				if resp.NextKey == "" {
+					return
+				}
+				pageToken = resp.NextKey
+			}
+		}
+
+		start := string(ordered.Encode(databaseName))
+		end := string(ordered.Encode(databaseName, ordered.Inf))
+		for ds := range rg.store.resources(databaseServerDatabaseNameIndex, start, end) {
```
Comment (espadolini) [inline lib/cache/database.go]: We can't be holding the cache lock while yielding into caller code that might block forever. Other implementations of iterators from the cache read a small-ish chunk in memory and yield from that, then continue iterating. In this case we should almost always end up with just one chunk with a small amount of items, but we should still chunk up the data to safeguard ourselves against fringe scenarios.

**Review by espadolini — CHANGES_REQUESTED** [review]

File: lib/auth/authclient/api.go (hunk 6 above)
Comment (wethreetrees) [reply lib/auth/authclient/api.go]: Renamed to `RangeDatabaseServersWithName`

File: lib/cache/database.go (hunk 7 above)
Comment (wethreetrees) [reply lib/cache/database.go]: Took a bit, but I think I have this handled. This now uses a `genericLister` with a custom `upstreamList` function that translates the new `databaseServerDatabaseNameIndex` keys to/from backend keys. This enables seamless continuation for fallback and cache recovery mid-iteration.

I also added a test suite to verify the happy path as well as the edge case continuation scenarios. Should be able to reuse the test helpers for App and Kubernetes servers in upcoming PRs.

File: lib/services/local/presence.go [hunk 8]
```diff
@@ -1055,6 +1055,33 @@ func (s *PresenceService) GetDatabaseServers(ctx context.Context, namespace stri
 	return servers, nil
 }
 
+// RangeDatabaseServersWithName returns an iterator over database proxy servers for a given database name.
+func (s *PresenceService) RangeDatabaseServersWithName(ctx context.Context, databaseName string) iter.Seq2[types.DatabaseServer, error] {
+	if databaseName == "" {
+		return func(yield func(types.DatabaseServer, error) bool) {
+			yield(nil, trace.BadParameter("missing database name"))
+		}
```
Comment (okraport) [inline lib/services/local/presence.go]: nit: we have helpers for this:
```suggestion
		return stream.Fail[types.DatabaseServer](trace.BadParameter("missing database name")))
```

File: lib/cache/database.go [hunk 9]
```diff
@@ -224,6 +237,74 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
+				StartKey:     backendKey,
+			})
+			if err != nil {
+				return nil, "", trace.Wrap(err)
+			}
+
+			var page []types.DatabaseServer
+			for _, r := range resp.Resources {
+				server, ok := r.(types.DatabaseServer)
+				if !ok {
```
Comment (okraport) [inline lib/cache/database.go:290]: We should log this.

File: lib/cache/database.go [hunk 10]
```diff
@@ -166,6 +169,15 @@ func (c *Cache) RangeDatabases(ctx context.Context, start, end string) iter.Seq2
 type databaseServerIndex string
 
 const databaseServerNameIndex databaseServerIndex = "name"
+const databaseServerDatabaseNameIndex databaseServerIndex = "database_name"
+
+func databaseServerByDatabaseNameKey(s types.DatabaseServer) string {
+	db := s.GetDatabase()
+	if db == nil {
+		return ""
+	}
```
Comment (okraport) [inline lib/cache/database.go:182]: Are we okay with silently dropping these entries? I know that technically this should never happen but in a case it does it may be better to store them in memory for debugging anyway. 

I think we should fix `s.GetDatabase().GetName()` to be nil safe and just do:
```go
string(ordered.Encode(s.GetDatabase().GetName(), s.GetHostID(), s.GetName()))
```

File: lib/cache/database.go [hunk 11]
```diff
@@ -224,6 +237,74 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
+// RangeDatabaseServersWithName returns an iterator over database proxy servers for a given database name.
+func (c *Cache) RangeDatabaseServersWithName(ctx context.Context, databaseName string) iter.Seq2[types.DatabaseServer, error] {
+	return func(yield func(types.DatabaseServer, error) bool) {
+		ctx, span := c.Tracer.Start(ctx, "cache/RangeDatabaseServersWithName")
+		defer span.End()
+
+		if databaseName == "" {
+			yield(nil, trace.BadParameter("missing database name"))
+			return
+		}
```
Comment (okraport) [inline lib/cache/database.go]: Is it worth checking this outside and just bailing out quickly with stream.Fail?

File: lib/cache/database.go (hunk 10 above)
Comment (wethreetrees) [reply lib/cache/database.go:182]: If I understand correctly, delete events use the collection's `headerTransform` to generate a resource and loop over all indexes to find the cache value to delete. The transform does not populate the database name, so on deletes it will always be `nil`.

I am currently returning an empty string in that case because we don't need the secondary index to find the cache value. The delete will find the value based on the primary index and also handles the index cleanup for ALL indexes for that ref value.

I can look into making `GetDatabase()` nil safe though. At the very least, maybe this needs a comment to explain the delete behavior?

Comment (okraport) [reply lib/cache/database.go:182]: This is a good point, apologies I forgot about the delete event. A comment sounds good because this isn't immediately obvious I think 👍

File: lib/cache/database.go (hunk 11 above)
Comment (espadolini) [reply lib/cache/database.go]: +1

File: lib/cache/database.go [hunk 12]
```diff
@@ -224,6 +237,74 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
+
+		upstreamListFn := func(ctx context.Context, pageSize int, startToken string) ([]types.DatabaseServer, string, error) {
+			// discard the databaseName, we don't need it to query the backend
+			var hostID, serverName string
+			_ = ordered.Decode([]byte(startToken), nil, &hostID, &serverName)
+			backendKey := ""
+			if hostID != "" {
+				backendKey = hostID + "/" + serverName
+			}
```
Comment (espadolini) [inline lib/cache/database.go:275]: This should be an explicit check against startToken being ~~empty~~ just the database name prefix, not a side effect of ordered.Decode erroring out.

File: lib/services/local/presence.go [hunk 13]
```diff
@@ -1055,6 +1055,33 @@ func (s *PresenceService) GetDatabaseServers(ctx context.Context, namespace stri
...
+		server, err := services.UnmarshalDatabaseServer(
+			item.Value,
+			services.WithExpires(item.Expires),
+			services.WithRevision(item.Revision),
+		)
+		if err != nil {
+			s.logger.WarnContext(ctx, "Failed to unmarshal database server", "key", item.Key, "error", err)
+			return nil, false
+		}
+		return server, server.GetDatabase().GetName() == databaseName
```
Comment (espadolini) [inline lib/services/local/presence.go:1078]: Is there a world in which we can decide that it's invalid for a database_server to not have the correct name, so we can avoid the unmarshal if the resource name doesn't match the database name that we want?

File: lib/cache/database.go [hunk 14]
```diff
@@ -224,6 +237,74 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
+		ctx, span := c.Tracer.Start(ctx, "cache/RangeDatabaseServersWithName")
+		defer span.End()
+
+		if databaseName == "" {
+			yield(nil, trace.BadParameter("missing database name"))
+			return
+		}
+
+		upstreamListFn := func(ctx context.Context, pageSize int, startToken string) ([]types.DatabaseServer, string, error) {
+			// discard the databaseName, we don't need it to query the backend
```
Comment (espadolini) [inline lib/cache/database.go]: I wonder if there's value in using the database name from the token for filtering, so this upstream list function becomes a pure function.

File: lib/services/local/presence.go (hunk 8 above)
Comment (wethreetrees) [reply lib/services/local/presence.go]: fixed in https://github.com/gravitational/teleport/pull/66484/commits/e4ace4c872d422ce968b99a4ba000cc913fdb8ab

File: lib/cache/database.go (hunk 11 above)
Comment (wethreetrees) [reply lib/cache/database.go]: fixed in https://github.com/gravitational/teleport/pull/66484/commits/e4ace4c872d422ce968b99a4ba000cc913fdb8ab

File: lib/cache/database.go (hunk 10 above)
Comment (wethreetrees) [reply lib/cache/database.go:182]: Added a comment in https://github.com/gravitational/teleport/pull/66484/commits/1a7eeafda59d496c09c5a8d2572164a84038e9aa

File: lib/cache/database.go (hunk 9 above)
Comment (wethreetrees) [reply lib/cache/database.go:290]: fixed in https://github.com/gravitational/teleport/pull/66484/commits/c0401eb73961053b160fb2187d3dc5da452c3598

File: lib/cache/database.go [hunk 15]
```diff
@@ -224,6 +237,74 @@ func (c *Cache) GetDatabaseServers(ctx context.Context, namespace string, opts .
...
+			var hostID, serverName string
+			_ = ordered.Decode([]byte(startToken), nil, &hostID, &serverName)
```
Comment (espadolini) [inline lib/cache/database.go]: Is there a reason to swallow the error rather than just returning it?

File: lib/cache/database.go (hunk 12 above)
Comment (espadolini) [reply lib/cache/database.go:275]: `ordered.DecodePrefix` should leave an empty byteslice after successfully decoding just the database name, which we can either compare to the expected database name and error if there's a discrepancy or take at face value and ignore the expected database name, then the rest is either empty (if this is the very first page) or must successfully decode into (host ID, name).

Comment (wethreetrees) [reply lib/cache/database.go:275]: That's a good plan. I spent some time thinking over options but never landed on processing twice, once for the database name and then for the hostID + serverName.

I'll check it out.

File: lib/cache/database.go (hunk 15 above)
Comment (wethreetrees) [reply lib/cache/database.go]: related to https://github.com/gravitational/teleport/pull/66484#discussion_r3267160186

If we hit the fallback on the first attempt, the startToken is just the database name, so an error was expected. In hindsight, I realize that it's not the best way to handle it 😅. I think this will resolve itself with the resolution of the other thread (i.e. better error handling).

File: lib/cache/database.go (hunk 14 above)
Comment (wethreetrees) [reply lib/cache/database.go]: It don't think be a pure function regardless because it uses the presence service, right?

I'm generally OK with using the db name from the token though. I originally captured it from the outer closure because nothing in the call path can modify it, which seemed simpler. Happy to switch if you'd prefer.

File: lib/cache/database.go (hunk 12 above)
Comment (wethreetrees) [reply lib/cache/database.go:275]: I split the decode, as discussed. It's still discarding the database name prefix and using the value from the outer scope but now has error handling for both decode steps.

Will update this dependent on the outcome of [this thread](https://github.com/gravitational/teleport/pull/66484#discussion_r3267224709).

File: lib/cache/database.go (hunk 14 above)
Comment (espadolini) [reply lib/cache/database.go]: True enough. Perhaps parsing and checking the database name against the expected one just to be sure we haven't hit some weird bug is the best way forward then.

File: lib/services/local/presence.go (hunk 13 above)
Comment (okraport) [reply lib/services/local/presence.go:1078]: In principle I agree with that world but we would have to codify that somewhere first. For the time being I would leave a comment here about this that it could be simplified if that decision is approved. Perhaps 153 could be a place to state this.

Comment (espadolini) [reply lib/services/local/presence.go:1078]: I don't think it's related to a "standard" at all, it'd be an optimization for this specific rpc related to the db_server resource given the fact that all legitimate db_server resources currently follow the naming scheme.

That being said, unmarshaling everything like this on a fallback is still an improvement over the current behavior of unmarshaling and collecting in a slice, and this being an optimization, it's something that we can do in the future, maybe if we decide to require names to match between the heartbeat resource and the inner resource at write time.

Comment (okraport) [reply lib/services/local/presence.go:1078]: My point was to the fact that currently we don't have an agreement in place to not violate that assumption in an unrelated change later on and introduce a regression here without really having any visibility into that. But what you said is true, so I'm happy either way.

File: lib/cache/database.go (hunk 14 above)
Comment (wethreetrees) [reply lib/cache/database.go]: Agreed, I'll get that updated.

Comment (wethreetrees) [reply lib/cache/database.go]: Updated in https://github.com/gravitational/teleport/pull/66484/commits/8785986da8730e16b0cbaa7cb422acd78ff64cf6

File: lib/cache/database.go (hunk 15 above)
Comment (wethreetrees) [reply lib/cache/database.go]: Resolved in https://github.com/gravitational/teleport/pull/66484/commits/afd88701c533ac43f09944f9b9790a22fdaf8f5c

File: lib/services/local/presence.go (hunk 13 above)
Comment (wethreetrees) [reply lib/services/local/presence.go:1078]: From what I can tell, the only way a database server could be created today with a mismatched database name would be through a custom integration. Is there any explicit support for something like that?

If we filter by the expected server name in the backend key before unmarshalling and then validated the inner database name, we would limit the affected path to MFA required checks when the cache is not yet ready for servers that:

- were created through a custom integration (with a server name that does not match it's database name)
- have no matching legitimate database service

FWIW, I'm also in support of just shipping the current iterative improvement for now.

Comment (okraport) [reply lib/services/local/presence.go:1078]: > From what I can tell, the only way a database server could be created today

The key operative word here is _today_. I think safest choice is to keep it as is and leave a comment here.

File: lib/cache/database.go (hunk 12 above)
Comment (wethreetrees) [reply lib/cache/database.go:275]: Added a compare for the token vs. expected database name in https://github.com/gravitational/teleport/pull/66484/commits/8785986da8730e16b0cbaa7cb422acd78ff64cf6

File: lib/services/local/presence.go (hunk 13 above)
Comment (wethreetrees) [reply lib/services/local/presence.go:1078]: Added a TODO in https://github.com/gravitational/teleport/pull/66484/commits/c6bf894949e3d0168389d0fa9aa0d374684f977b

**Review by okraport — APPROVED** [review]
The latest fallback ranger looks good.

**Review by espadolini — APPROVED** [review]
